stream/stream.go

```diff
@@ -4,229 +4,230 @@ package stream
 
 import (
 	"context"
-	"encoding/binary"
 	"encoding/json"
 	"errors"
 	"fmt"
+	"sync"
 	"time"
 
 	"github.com/lthibault/log"
 	"github.com/prometheus/client_golang/prometheus"
 
+	"github.com/blocknative/dreamboat/stream/transport"
 	"github.com/blocknative/dreamboat/structs"
 	"github.com/blocknative/dreamboat/structs/forks/bellatrix"
 	"github.com/blocknative/dreamboat/structs/forks/capella"
 )
 
-var (
+const (
 	CacheTopic         = "/block/cache"
 	BidTopic           = "/block/bid"
 	SlotDeliveredTopic = "/slot/delivered"
 )
 
-type Pubsub interface {
-	Publish(context.Context, string, []byte) error
-	Subscribe(context.Context, string) chan []byte
-}
-
-type StreamConfig struct {
-	Logger          log.Logger
-	ID              string
-	PubsubTopic     string // pubsub topic name for block submissions
-	StreamQueueSize int
+type PubSub interface {
+	Publish(context.Context, transport.Message) error
+	Subscribe(context.Context) transport.Subscription
 }
 
 type State interface {
 	ForkVersion(epoch structs.Slot) structs.ForkVersion
 	HeadSlot() structs.Slot
 }
 
+type Metrics interface {
+	Register(prometheus.Collector) error
+}
+
 type Client struct {
-	Pubsub Pubsub
+	once sync.Once
 
-	builderBidIn     chan []byte
-	builderBidOut    chan structs.BuilderBidExtended
-	cacheIn          chan []byte
-	cacheOut         chan structs.BlockAndTraceExtended
-	slotDeliveredIn  chan []byte
-	slotDeliveredOut chan uint64
+	Logger                log.Logger
+	State                 State
+	Bids, Cache           PubSub
+	QueueSize, NumWorkers int
 
-	Config StreamConfig
-	Logger log.Logger
+	Metrics Metrics
+	m       streamMetrics
 
-	m StreamMetrics
+	ctx    context.Context
+	cancel context.CancelFunc
 
-	//slotDelivered chan structs.Slot
+	builderBidOut chan structs.BuilderBidExtended
+	cacheOut      chan structs.BlockAndTraceExtended
 
 	st State
 }
 
-func NewClient(ps Pubsub, st State, cfg StreamConfig) *Client {
-	s := Client{
-		Pubsub: ps,
-		st:     st,
-
-		builderBidIn:     make(chan []byte, cfg.StreamQueueSize),
-		builderBidOut:    make(chan structs.BuilderBidExtended, cfg.StreamQueueSize),
-		cacheIn:          make(chan []byte, cfg.StreamQueueSize),
-		cacheOut:         make(chan structs.BlockAndTraceExtended, cfg.StreamQueueSize),
-		slotDeliveredIn:  make(chan []byte, cfg.StreamQueueSize),
-		slotDeliveredOut: make(chan uint64, cfg.StreamQueueSize),
+func (c *Client) init() {
+	c.once.Do(func() {
+		c.initMetrics()
 
-		Config: cfg,
-		Logger: cfg.Logger.WithField("subService", "stream").WithField("type", "redis"),
-	}
+		if c.Logger == nil {
+			c.Logger = log.New()
+		}
 
-	s.initMetrics()
+		c.builderBidOut = make(chan structs.BuilderBidExtended, c.QueueSize)
+		c.cacheOut = make(chan structs.BlockAndTraceExtended, c.QueueSize)
 
-	return &s
+		c.ctx, c.cancel = context.WithCancel(context.Background())
+		go c.subscribe(c.Bids, c.handleBid)
+		go c.subscribe(c.Cache, c.handleCache)
```

Comment:
I'm not quite sure I follow; do you want to explicitly add subscriptions at the `Client` call site (i.e. in `main.go`)?

Reply:
I'm trying to avoid "callback function pattern" here :) 
And it may make sense to just hide `chan builderBidOut` inside another structure - `BidSubscription` in the example above. This way BlockCache method can become a BidSubscription's and can be potentially removed from the Client. 
Maybe `BidSubscriptionHandler` would be a better name?
So from the Client structure perspective, the interface would be just: 
`type SubscribeFor interface {
     Handle(m Message) error
}`

Reply:
>And it may make sense to just hide chan builderBidOut inside another structure

This would require each transport implementation to manage the lifecycle of a channel.  In addition to not being very DRY, it also exposes us to leaky abstractions, since each can now have different queuing/concurrency strategies.  So this feels like it would be going in the wrong direction.

I think callbacks are generally fine, provided they're not abused.  It's really no different from providing an interface.